File: online-judges-and-contests/beecrowd/2082v.go
package main

import (
	"container/list"
	"fmt"
)

const INF = 1e9

type Aresta struct {
	destino   int
	capacidade int
}

type Grafo struct {
	numVertices  int
	adjacencias [][]Aresta
}

func novoGrafo(n int) *Grafo {
	return &Grafo{
		numVertices:  n,
		adjacencias: make([][]Aresta, n),
	}
}

func (g *Grafo) adicionarAresta(verticeOrigem, verticeDestino, capacidade int) {
	g.adjacencias[verticeOrigem] = append(g.adjacencias[verticeOrigem], Aresta{verticeDestino, capacidade})
	g.adjacencias[verticeDestino] = append(g.adjacencias[verticeDestino], Aresta{verticeOrigem, capacidade})
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func algFordFulkerson(grafo *Grafo, fonte, escoadouro int) int {
	n := grafo.numVertices
	matrizResidual := make([][]int, n)
	for i := range matrizResidual {
		matrizResidual[i] = make([]int, n)
	}
	paiVertices := make([]int, n)
	fluxoMax := 0

	for {
		for i := range paiVertices {
			paiVertices[i] = -1
		}

		fila := list.New()
		fila.PushBack([2]int{fonte, INF})

		for fila.Len() > 0 {
			elem := fila.Front()
			fila.Remove(elem)
			atual := elem.Value.([2]int)[0].(int)
			fluxo := elem.Value.([2]int)[1].(int)

			for _, aresta := range grafo.adjacencias[atual] {
				prox := aresta.destino
				if paiVertices[prox] == -1 && aresta.capacidade-matrizResidual[atual][prox] > 0 {
					paiVertices[prox] = atual
					novoFluxo := min(fluxo, aresta.capacidade-matrizResidual[atual][prox])
					if prox == escoadouro {
						fluxoMax += novoFluxo
						esc := prox
						for esc != fonte {
							ant := paiVertices[esc]
							matrizResidual[ant][esc] += novoFluxo
							matrizResidual[esc][ant] -= novoFluxo
							esc = ant
						}
						break
					}
					fila.PushBack([2]int{prox, novoFluxo})
				}
			}
		}

		if paiVertices[escoadouro] == -1 {
			break
		}
	}

	return fluxoMax
}

func main() {
	var T int
	fmt.Scan(&T)

	for t := 0; t < T; t++ {
		var n, m int
		fmt.Scan(&n, &m)

		grafo := novoGrafo(n)

		for i := 0; i < m; i++ {
			var u, v, c int
			fmt.Scan(&u, &v, &c)
			grafo.adicionarAresta(u-1, v-1, c)
		}

		fonte := 0
		minCut := INF

		for u := 1; u < n; u++ {
			matrizResidualGrafo := *grafo
			fluxoMax := algFordFulkerson(&matrizResidualGrafo, fonte, u)
			minCut = min(minCut, fluxoMax)
		}

		fmt.Println(minCut)
	}
}

File: online-judges-and-contests/beecrowd/2082v.rs
use std::collections::{VecDeque, HashMap};

const INF: i32 = 1e9 as i32;

#[derive(Clone)]
struct Aresta {
    destino: usize,
    capacidade: i32,
}

struct Grafo {
    num_vertices: usize,
    adjacencias: Vec<Vec<Aresta>>,
}

impl Grafo {
    fn new(n: usize) -> Self {
        Self {
            num_vertices: n,
            adjacencias: vec![vec![]; n],
        }
    }

    fn adicionar_aresta(&mut self, vertice_origem: usize, vertice_destino: usize, capacidade: i32) {
        self.adjacencias[vertice_origem].push(Aresta {
            destino: vertice_destino,
            capacidade,
        });
        self.adjacencias[vertice_destino].push(Aresta {
            destino: vertice_origem,
            capacidade,
        });
    }
}

fn alg_ford_fulkerson(grafo: &mut Grafo, fonte: usize, escoadouro: usize) -> i32 {
    let n = grafo.num_vertices;
    let mut matriz_residual = vec![vec![0; n]; n];
    let mut pai_vertices = vec![-1; n];
    let mut fluxo_max = 0;

    while {
        pai_vertices.iter_mut().for_each(|v| *v = -1);
        let mut fila = VecDeque::new();
        fila.push_back((fonte, INF));
        let mut found_path = false;

        while let Some((atual, fluxo)) = fila.pop_front() {
            for aresta in &grafo.adjacencias[atual] {
                let prox = aresta.destino;
                if pai_vertices[prox] == -1 && aresta.capacidade - matriz_residual[atual][prox] > 0 {
                    pai_vertices[prox] = atual as i32;
                    let novo_fluxo = std::cmp::min(fluxo, aresta.capacidade - matriz_residual[atual][prox]);
                    if prox == escoadouro {
                        fluxo_max += novo_fluxo;
                        let mut esc = prox as i32;
                        while esc != fonte as i32 {
                            let ant = pai_vertices[esc as usize];
                            matriz_residual[ant as usize][esc as usize] += novo_fluxo;
                            matriz_residual[esc as usize][ant as usize] -= novo_fluxo;
                            esc = ant;
                        }
                        found_path = true;
                        break;
                    }
                    fila.push_back((prox, novo_fluxo));
                }
            }
            if found_path {
                break;
            }
        }

        found_path
    } {}

    fluxo_max
}

fn main() {
    let mut input = String::new();
    std::io::stdin().read_line(&mut input).unwrap();
    let t: usize = input.trim().parse().unwrap();

    for _ in 0..t {
        let mut input = String::new();
        std::io::stdin().read_line(&mut input).unwrap();
        let mut iter = input.split_whitespace();
        let n: usize = iter.next().unwrap().parse().unwrap();
        let m: usize = iter.next().unwrap().parse().unwrap();

        let mut grafo = Grafo::new(n);

        for _ in 0..m {
            let mut input = String::new();
            std::io::stdin().read_line(&mut input).unwrap();
            let mut iter = input.split_whitespace();
            let u: usize = iter.next().unwrap().parse().unwrap();
            let v: usize = iter.next().unwrap().parse().unwrap();
            let c: i32 = iter.next().unwrap().parse().unwrap();
            grafo.adicionar_aresta(u - 1, v - 1, c);
        }

        let fonte = 0;
        let mut min_cut = INF;

        for u in 1..n {
            let mut grafo_residual = grafo.clone();
            let fluxo_max = alg_ford_fulkerson(&mut grafo_residual, fonte, u);
            min_cut = min_cut.min(fluxo_max);
        }

        println!("{}", min_cut);
    }
}
